### PR TITLE
Authority Discovery to FRAME v2

### PR DESCRIPTION
**Pull request by ferrell-code**
migrated the pallet to new macro :). I believe there is a minor issue with the macro in genesis config. The assimilate_storage function needs the Config trait for genesis builder, but when the genesis config doesn't have the Config trait it won't work. Delete my manual impl of assimilate_storage and try to run the tests and you will see what I mean

### CODE REVIEW

File: frame/authority-discovery/src/lib.rs [hunk 1]
```diff
@@ -15,45 +15,87 @@
...
-		NextKeys get(fn next_keys): Vec<AuthorityId>;
-	}
-	add_extra_genesis {
-		config(keys): Vec<AuthorityId>;
-		build(|config| Module::<T>::initialize_keys(&config.keys))
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::{
+		pallet_prelude::*,
+	};
```
Comment (thiolliere) [inline frame/authority-discovery/src/lib.rs:36]:
```suggestion
	use frame_support::pallet_prelude::*;
```

File: frame/authority-discovery/src/lib.rs [hunk 2]
```diff
@@ -15,45 +15,87 @@
...
+		_,
+		Vec<AuthorityId>,
+		ValueQuery,
+	>;
+	
+	#[pallet::storage]
+	#[pallet::getter(fn next_keys)]
+	/// Keys of the next authority set.
+	pub(super) type NextKeys<T: Config> = StorageValue<
+		_,
```
Comment (thiolliere) [inline frame/authority-discovery/src/lib.rs:61]:
```suggestion
		_,
```

File: frame/authority-discovery/src/lib.rs [hunk 3]
```diff
@@ -113,6 +155,16 @@ impl<T: Config> OneSessionHandler<T::AccountId> for Module<T> {
 	}
 }
 
+impl GenesisConfig {
+	/// Direct implementation of `GenesisBuild::assimilate_storage`.	
+	pub	fn assimilate_storage<T: Config>(
```
Comment (thiolliere) [inline frame/authority-discovery/src/lib.rs:160]:
```suggestion
	#[deprecated(note = "use [`GenesisBuild::assimilate_storage`] instead")]
	pub fn assimilate_storage<T: Config>(
```

File: frame/authority-discovery/src/lib.rs [hunk 4]
```diff
@@ -113,6 +155,16 @@ impl<T: Config> OneSessionHandler<T::AccountId> for Module<T> {
 	}
 }
 
+impl GenesisConfig {
+	/// Direct implementation of `GenesisBuild::assimilate_storage`.
```
Comment (thiolliere) [inline frame/authority-discovery/src/lib.rs:159]:
```suggestion
	/// Direct implementation of `GenesisBuild::assimilate_storage`.
```

**Review by thiolliere — APPROVED** [review]
## ⚠️ Breaking Change ⚠️ 

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the `AuthorityDiscovery` pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the `AuthorityDiscovery` pallet.

name of the pallet in both kusama, polkadot and westend are: `AuthorityDiscovery` thus no need for migration.

Looks good.

File: frame/authority-discovery/src/lib.rs [hunk 5]
```diff
@@ -15,45 +15,87 @@
...
-//! This module is used by the `client/authority-discovery` and by polkadot's parachain logic
+//! This pallet is used by the `client/authority-discovery` and by polkadot's parachain logic
 //! to retrieve the current and the next set of authorities.
 
 // Ensure we're `no_std` when compiling for Wasm.
 #![cfg_attr(not(feature = "std"), no_std)]
 
 use sp_std::prelude::*;
-use frame_support::{decl_module, decl_storage, traits::OneSessionHandler};
+use frame_support::traits::{OneSessionHandler, GenesisBuild};
```
Comment (thiolliere) [inline frame/authority-discovery/src/lib.rs:27]:
```suggestion
use frame_support::traits::OneSessionHandler;
#[cfg(feature = "std")]
use frame_support::traits::GenesisBuild;
```

File: frame/authority-discovery/src/lib.rs [hunk 6]
```diff
@@ -113,6 +155,16 @@ impl<T: Config> OneSessionHandler<T::AccountId> for Module<T> {
 	}
 }
 
+impl GenesisConfig {
```
Comment (thiolliere) [inline frame/authority-discovery/src/lib.rs:158]:
```suggestion
#[cfg(feature = "std")]
impl GenesisConfig {
```

**Comment by shawntabrizi** [comment]
@ferrell-code got a compiler error here:

```
error: use of deprecated associated function `<impl pallet::GenesisConfig>::assimilate_storage`: use [`GenesisBuild::assimilate_storage`] instead
   --> frame/authority-discovery/src/lib.rs:316:4
    |
316 |         .assimilate_storage::<Test>(&mut t)
    |          ^^^^^^^^^^^^^^^^^^
    |
    = note: `-D deprecated` implied by `-D warnings`
```

Are you able to resolve it?

File: frame/authority-discovery/src/lib.rs [hunk 7]
```diff
@@ -113,6 +155,18 @@ impl<T: Config> OneSessionHandler<T::AccountId> for Module<T> {
 	}
 }
 
+#[cfg(feature = "std")]
+impl GenesisConfig {
+	/// Direct implementation of `GenesisBuild::assimilate_storage`.
+	#[deprecated(note = "use [`GenesisBuild::assimilate_storage`] instead")]
```
Comment (thiolliere) [inline frame/authority-discovery/src/lib.rs:161]:
```suggestion
```

**Comment by thiolliere** [comment]
> @ferrell-code got a compiler error here:
> 
> ```
> error: use of deprecated associated function `<impl pallet::GenesisConfig>::assimilate_storage`: use [`GenesisBuild::assimilate_storage`] instead
>    --> frame/authority-discovery/src/lib.rs:316:4
>     |
> 316 |         .assimilate_storage::<Test>(&mut t)
>     |          ^^^^^^^^^^^^^^^^^^
>     |
>     = note: `-D deprecated` implied by `-D warnings`
> ```
> 
> Are you able to resolve it?

Sorry I introduced by deprecated the assimilate_storage.

I think we can remove or deprecate the implementation though, but we can do that in later PR it is fine to me.

**Comment by ferrell-code** [comment]
Cool I get it, assimilate_storage method is the old way of doing it. I can update the tests to remove the implementation (or deprecate it if necessary), idk if you'd rather it be in this PR or open another one

**Comment by thiolliere** [comment]
> Cool I get it, assimilate_storage method is the old way of doing it. I can update the tests to remove the implementation (or deprecate it if necessary), idk if you'd rather it be in this PR or open another one

this pattern of reexposing manually the old method instead of using GenesisBuild has been used in other crates as well. So we will need to clean it in the future anyway.

You can do for this pallet in this PR if you want, other in a follow-up, or we'll do once we deprecate all of those.

If you want to merge as-is could you merge master to make CI happy ?

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Comment by thiolliere** [comment]
thank you @ferrell-code

**Pull request merged**